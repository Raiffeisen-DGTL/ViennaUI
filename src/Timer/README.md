# Timer

Комопнент таймера, обновляющий свое содержимое с заданным интервалом.

## Импорт

```
import { Timer, useInterval } from 'vienna-ui';
```

## Свойства / Props

| Prop           | Type                                                             | Default | Description            |
| -------------- | ---------------------------------------------------------------- | ------- | ---------------------- |
| start          | number \| undefined                                              | 0       | Начальное значения     |
| step           | number \| undefined                                              | 1       |
| stop           | number \| undefined                                              | false   | Конечное значение      |
| delay          | number \| undefined                                              | 1000    | Интервал обновления    |
| countdown      | boolean \| undefined                                             | false   |
| allowNegatives | boolean \| undefined                                             | false   | Отрицательные значения |
| onChange       | ((count: number, id?: string \| undefined) => void) \| undefined | false   |
| onStop         | ((count: number, id?: string \| undefined) => void) \| undefined | false   |

## Использование

`<Timer />`

## Использование как render function

По умолчанию таймер будет отображать внутри себя текущую итерацию в виде текста без всяких обёрток. В дополнение к этому, в содержимое компонента можно передать render-функцию, которая в аргументах будет принимать номер итерации и `id` компонента и возвращать содержимое для компонента.

```
<Timer id='simple-timer'>
    {(tick, id) => (
        <div style={{ color: '#808185' }}>
            <div>Timer id: '{id}'</div>
            <div>Current tick: {tick}</div>
        </div>
    )}
</Timer>
```

## Интервал обновления

##### Свойство `delay`

Интервал обновления можно колнтролировать с помощью параметра `delay`. Интервал укаазыватеся в милисекунщах, значение по умолчанию равно 1000 (1с). Установка `delay` в null остановит таймер.

```
<Groups design='vertical'>
    <Timer delay={500} />
    <Timer />
</Groups>
```

## Начальное и конечное значения

##### Свойство `start`

##### Свойство `stop`

С помощью парасемтров `start` / `stop` можно установить начальное и конечное значения таймера.

`<Timer start={100} stop={110} />`

## Шаг

##### Свойство `step`

Шагом таймера можно управлять с помощью параметра `step`.

`<Timer step={100} />`

## Режим обратного отсчета

##### Свойство `countdown`

С помощью флага `countdown` можно переключить таймер в режим обратного отсчёта.

`<Timer start={50} countdown />`

## Отрицательные значения

##### Свойство `allowNegatives`

По умолчанию таймер обратного отсчета остановится при достижении 0. Разрешить переход в отрицательное множество, можно с помощью флага `allowNegatives`.

`<Timer start={10} countdown allowNegatives />`

## Date and Time

> Так как под капотом используется setInterval, все его ограничения распросраняются на работу компонента. В частности возможный браузерный тротлинг при работе в неактивной вкладке.
>
> Поэтому, для отображения текущего времени, нельзя полагаться на номер итерации таймера. Необходимо на каждой итерации получать текущее время с помощью объекта `Date`

`` <Timer>{() => `Current time: ${new Date().toLocaleTimeString()}`}</Timer> ``

## useInterval

В дополнение к самому компоненту Дизайн-Система экспрортирует хук `useInterval`, являющийся оберткой над нативным `setInterval`.

Так же как и в компоненте, при передаче `null` в `delay` таймер остановится.

```
{() => {
    const [count, update] = React.useState(0);
    const [stoped, stop] = React.useState(false);
    useInterval(
        () => {
            update(count + 1);
        },
        stoped ? null : 1000
    );
    return (
        <Groups>
            <div>{count}</div>
            <Button size='s' onClick={() => stop(!stoped)}>
                Start / Stop timer
            </Button>
        </Groups>
    );
}}
```

## Event handlers

##### Свойство `onChange`

##### Свойство `onStop`

Обратботчики событий таймера иимеют единую сигнатуру, первым параметром передается текущая итерация таймера, вторым – идентификатор элемента.

Обработчик `onChange` вызыватеся в каждую итерацию таймера.

`` <Timer id='on-change' onChange={(tick, id) => console.log(`tick: ${tick}, timer id: ${id}`)} /> ``

Если для таймера задано условие остановки (дефолтный 0 для таймеров обратного отсчёта или явно переданый параметр `stop`), то при остановки таймера будет также вызыван обработчик `onStop`.

`` <Timer id='on-stop' stop={10} onStop={(tick, id) => console.log(`tick: ${tick}, timer id: ${id}`)} /> ``
