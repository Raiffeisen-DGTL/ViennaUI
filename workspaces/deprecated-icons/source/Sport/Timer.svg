<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg"><path d="M17.852 8.563h-.002a8 8 0 11-11.7 0L4.294 6.707l1.414-1.418 1.98 1.98a7.95 7.95 0 013.313-1.2V4h-1V2h4v2h-1v2.069a7.95 7.95 0 013.313 1.204l1.98-1.98 1.414 1.414-1.856 1.856zM12.001 20a6 6 0 100-12 6 6 0 000 12zm-1-11h2v5h-2V9z"/></svg>
