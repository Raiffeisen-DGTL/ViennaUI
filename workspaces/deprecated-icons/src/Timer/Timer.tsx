import React, { SVGAttributes } from 'react';

export interface TimerProps extends SVGAttributes<SVGElement> {
    [key: string]: any;
    size?: 'xs' | 's' | 'm' | 'l' | 'xl' | number;
    color?: string;
}

const sizes = { xs: 12, s: 16, m: 20, l: 24, xl: 32 };
export const Timer: React.FC<TimerProps> = (props): JSX.Element => {
    const { color = 'currentColor', size = 'm', ...attrs } = props;
    const d = sizes[size] || size;

    return (
        <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24' width={d} height={d} fill={color} {...attrs}>
            <defs />
            <path d='M17.852 8.563h-.002a8 8 0 11-11.7 0L4.294 6.707l1.414-1.418 1.98 1.98a7.95 7.95 0 013.313-1.2V4h-1V2h4v2h-1v2.069a7.95 7.95 0 013.313 1.204l1.98-1.98 1.414 1.414-1.856 1.856zM12.001 20a6 6 0 100-12 6 6 0 000 12zm-1-11h2v5h-2V9z' />
        </svg>
    );
};

Timer.defaultProps = {
    size: 'm',
    color: 'currentColor',
};

Timer.displayName = 'Timer';
export default Timer;
