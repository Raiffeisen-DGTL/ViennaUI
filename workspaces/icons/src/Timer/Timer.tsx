import React, { SVGAttributes } from 'react';

export interface TimerProps extends SVGAttributes<SVGElement> {
    [key: string]: any;
    size?: 'xs' | 's' | 'm' | 'l' | 'xl' | number;
    color?: string;
}

const sizes = { xs: 12, s: 16, m: 20, l: 24, xl: 32 };
export const Timer: React.FC<TimerProps> = (props): JSX.Element => {
    const { color = 'currentColor', size = 'm', ...attrs } = props;
    const d = sizes[size] || size;

    return (
        <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24' width={d} height={d} fill={color} {...attrs}>
            <defs />
            <path d='M11 10v5h2v-5h-2z' />
            <path
                fillRule='evenodd'
                d='M13 4h2V2H9v2h2v1.558a8.45 8.45 0 00-3.681 1.346L5.707 5.293 4.293 6.707 5.786 8.2a8.5 8.5 0 1012.428 0l1.493-1.493-1.414-1.414-1.612 1.611A8.45 8.45 0 0013 5.558V4zM5.5 14a6.5 6.5 0 1113 0 6.5 6.5 0 01-13 0z'
                clipRule='evenodd'
            />
        </svg>
    );
};

Timer.defaultProps = {
    size: 'm',
    color: 'currentColor',
};

Timer.displayName = 'Timer';
export default Timer;
