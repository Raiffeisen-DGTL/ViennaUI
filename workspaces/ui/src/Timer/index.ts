export * from './Timer';
export * from './useInterval';
