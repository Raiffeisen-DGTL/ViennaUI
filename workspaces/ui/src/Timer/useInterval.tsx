/** Â© Dan Abramov https://overreacted.io/making-setinterval-declarative-with-react-hooks/ */
import { useEffect, useRef } from 'react';

export function useInterval(callback, delay) {
    const savedCallback = useRef(callback);

    // Remember the latest callback.
    useEffect(() => {
        savedCallback.current = callback;
    }, [callback]);

    // Set up the interval.
    useEffect(() => {
        function tick() {
            savedCallback.current();
        }

        if (delay !== null) {
            const id = setInterval(tick, delay);
            return () => clearInterval(id);
        }

        return undefined;
    }, [delay]);
}

export default useInterval;
