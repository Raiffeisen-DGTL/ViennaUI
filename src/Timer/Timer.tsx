import { FC, ReactNode, useState } from 'react';
import { useInterval } from './useInterval';

export interface TimerProps {
    children?: ReactNode;
    id?: string;
    start?: number;
    stop?: number;
    delay?: number;
    step?: number;
    countdown?: boolean;
    allowNegatives?: boolean;
    onChange?: (count: number, id?: string) => void;
    onStop?: (count: number, id?: string) => void;
}

export const Timer: FC<TimerProps> = ({
    children,
    id,
    stop,
    delay = 1000,
    countdown,
    allowNegatives,
    onChange,
    onStop,
    start = 0,
    step = 1,
}: TimerProps) => {
    const [count, setCount] = useState(start);

    const stopNow = (countdown && !allowNegatives && count === 0) || (stop && stop === count);

    useInterval(
        () => {
            setCount(countdown ? (c) => c - step : (c) => c + step);
            onChange?.(count, id);
        },
        stopNow ? undefined : delay
    );

    if (stopNow && onStop) {
        onStop(count, id);
    }

    // @ts-ignore
    return children && typeof children === 'function' ? children(count, id) : count;
};

Timer.displayName = 'Timer';
