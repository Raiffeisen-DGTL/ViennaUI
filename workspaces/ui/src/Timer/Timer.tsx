import React, { useState } from 'react';
import { useInterval } from './useInterval';

interface TimerProps {
    children?: React.ReactNode;
    id?: string;
    start?: number;
    stop?: number;
    delay?: number;
    step?: number;
    countdown?: boolean;
    allowNegatives?: boolean;
    onChange?: (count: number, id?: string) => void;
    onStop?: (count: number, id?: string) => void;
}

export const Timer: React.FC<TimerProps> = (props: TimerProps) => {
    const { children, id, stop, delay, countdown, allowNegatives, onChange, onStop, start = 0, step = 1 } = props;
    const [count, setCount] = useState(start);

    const stopNow = (countdown && !allowNegatives && count === 0) || (stop && stop === count);

    useInterval(
        () => {
            setCount(countdown ? (c) => c - step : (c) => c + step);

            if (onChange) {
                onChange(count, id);
            }
        },
        stopNow ? null : delay
    );

    if (stopNow && onStop) {
        onStop(count, id);
    }

    return children && typeof children === 'function' ? children(count, id) : count;
};

Timer.defaultProps = {
    delay: 1000,
    start: 0,
    step: 1,
};

Timer.displayName = 'Timer';
export default Timer;
