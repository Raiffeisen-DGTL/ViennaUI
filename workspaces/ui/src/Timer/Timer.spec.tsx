import React from 'react';
import { act } from 'react-dom/test-utils';
import { Timer } from './Timer';

test('Timer', () => {
    const component = <Timer />;
    const snap = snapshot.mount(component);
    expect(snap).toMatchSnapshot();
    jest.clearAllTimers();
});

test('Timer w/ render function', () => {
    const component = (
        <Timer id='timer'>
            {(tick, id) => (
                <div id='test'>
                    Timer id: '{id}', tick: {tick}
                </div>
            )}
        </Timer>
    );

    const snap = snapshot.mount(component);

    expect(snap).toMatchSnapshot();
    jest.clearAllTimers();
});

test('Timer w/ start', () => {
    const component = (
        <Timer id='timer' start={10}>
            {(tick, id) => (
                <div id='test'>
                    Timer id: '{id}', tick: {tick}
                </div>
            )}
        </Timer>
    );

    const snap = snapshot.mount(component);

    expect(snap).toMatchSnapshot();
    jest.clearAllTimers();
});

jest.useFakeTimers();
test('Timer w/ handlers', () => {
    const onChange = jest.fn();
    const onStop = jest.fn();

    const component = <Timer id='timer' onChange={onChange} onStop={onStop} stop={10} />;

    snapshot.mount(component);

    act(() => {
        jest.advanceTimersByTime(10000);
    });

    expect(onChange).toBeCalled();
    expect(onChange).toHaveBeenCalledTimes(10);

    expect(onStop).toBeCalled();
    expect(onStop).toHaveBeenCalledTimes(1);
    expect(onStop).toHaveBeenCalledWith(10, 'timer');

    jest.clearAllTimers();
});

jest.useFakeTimers();
test('Timer w/ delay', () => {
    const onChange = jest.fn();
    const onStop = jest.fn();

    const component = <Timer id='timer' onChange={onChange} onStop={onStop} stop={20} delay={500} />;

    snapshot.mount(component);

    act(() => {
        jest.advanceTimersByTime(10000);
    });

    expect(onChange).toBeCalled();
    expect(onChange).toHaveBeenCalledTimes(20);

    expect(onStop).toBeCalled();
    expect(onStop).toHaveBeenCalledTimes(1);
    expect(onStop).toHaveBeenCalledWith(20, 'timer');

    jest.clearAllTimers();
});

jest.useFakeTimers();
test('Timer w/ step', () => {
    const onChange = jest.fn();
    const onStop = jest.fn();

    const component = <Timer id='timer' onChange={onChange} onStop={onStop} stop={1000} step={100} />;

    snapshot.mount(component);

    act(() => {
        jest.advanceTimersByTime(10000);
    });

    expect(onChange).toBeCalled();
    expect(onChange).toHaveBeenCalledTimes(10);

    expect(onStop).toBeCalled();
    expect(onStop).toHaveBeenCalledTimes(1);
    expect(onStop).toHaveBeenCalledWith(1000, 'timer');

    jest.clearAllTimers();
});

jest.useFakeTimers();
test('Timer w/ allowNegatives', () => {
    const onChange = jest.fn();
    const onStop = jest.fn();

    const component = (
        <Timer id='timer' onChange={onChange} onStop={onStop} start={10} stop={-10} countdown allowNegatives />
    );

    snapshot.mount(component);

    act(() => {
        jest.advanceTimersByTime(20000);
    });

    expect(onChange).toBeCalled();
    expect(onChange).toHaveBeenCalledTimes(20);

    expect(onStop).toBeCalled();
    expect(onStop).toHaveBeenCalledTimes(1);
    expect(onStop).toHaveBeenCalledWith(-10, 'timer');

    jest.clearAllTimers();
});
