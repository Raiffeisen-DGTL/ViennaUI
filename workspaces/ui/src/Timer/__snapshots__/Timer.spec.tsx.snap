// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer 1`] = `
<Timer
  delay={1000}
  start={0}
  step={1}
>
  0
</Timer>
`;

exports[`Timer w/ render function 1`] = `
<Timer
  delay={1000}
  id="timer"
  start={0}
  step={1}
>
  <div
    id="test"
  >
    Timer id: '
    timer
    ', tick: 
    0
  </div>
</Timer>
`;

exports[`Timer w/ start 1`] = `
<Timer
  delay={1000}
  id="timer"
  start={10}
  step={1}
>
  <div
    id="test"
  >
    Timer id: '
    timer
    ', tick: 
    10
  </div>
</Timer>
`;
